### DISCUSSION TITLE
Go to definition (gd) stops working

### DISCUSSION THREAD

**Discussion by akoskm** [LazyVim/LazyVim#2363] [Q&A]
Hey 👋 

I'm working on a fairly large TypeScript project and trying to switch from VSCode. Currently, the only thing blocking me is the unreliability of **gd**. I'd often like to jump to the definition of an imported React component or type. Sometimes - usually around when I start LazyVim it works. Up until a point where it doesn't work anymore (nothing happens when I press **gd**).

LspRestart/LspStart/LspStop doesn't fix this.

When I put the cursor over the word I want to **gd** on, and manually type `:Telescope` and filter for `lsp_definitions` and hit enter, the dialog disappears, but the editor doesn't jump to the definition.

This works in VSCode.

Here's the list of my loaded plugins:

![image](https://github.com/LazyVim/LazyVim/assets/3111627/169142de-bf0b-4652-9176-c4ac89d60d1d)

`LspInfo`

![image](https://github.com/LazyVim/LazyVim/assets/3111627/2cc26c0d-35b4-41b4-aa12-70dc32e8ef10)

and `Mason`

![image](https://github.com/LazyVim/LazyVim/assets/3111627/dd21c8f5-a2c9-413e-bbeb-95240a73011e)

**Comment by jyuan0** [comment]
I'm doubtful I'll be able to help you much, but are there any clues/errors in your LSP log?

**Reply by akoskm** [reply]
Not really. I created a new log file and restarted neovim. When I opened the first typescript file I got the message Initializing JS/TS language features... and this is all I have in the fresh logs:

```
[START][2024-01-18 06:31:47] LSP logging initiated
[WARN][2024-01-18 06:31:47] ...lsp/handlers.lua:137^I"The language server eslint triggers a registerCapability handler despite dynamicRegistration set to false. Report upstream, this warning is harmless"
[WARN][2024-01-18 06:31:49] ...lsp/handlers.lua:137^I"The language server eslint triggers a registerCapability handler despite dynamicRegistration set to false. Report upstream, this warning is harmless"
[WARN][2024-01-18 06:31:49] ...lsp/handlers.lua:137^I"The language server eslint triggers a registerCapability handler despite dynamicRegistration set to false. Report upstream, this warning is harmless"
[ERROR][2024-01-18 06:32:42] ...lsp/handlers.lua:535^I"[lspserver] [tsclient] [tsserver] Exited. Code: null. Signal: SIGABRT"
```

Now even when I type the LspStart command, it doesn't do anything. I guess LazyVim thinks it has already started, but looking at the logs, it crashed.


What are the correct packages to install to get ts support?

Should I have both: tsserver from mason and null-ls installed?

**Reply by jyuan0** [reply]
> What are the correct packages to install to get ts support?

Not sure, but you can check what LazyVim sets up: https://github.com/LazyVim/LazyVim/blob/main/lua/lazyvim/plugins/extras/lang/typescript.lua

> Should I have both: tsserver from mason and null-ls installed?

That's fine. Mason is for installing tools, and null-ls/none-ls provides LSP functionality.

I wonder if `LspRestart` makes a difference, but besides increasing log verbosity I don't really have any other ideas at the moment.

**Reply by akoskm** [reply]
Thanks for that link, I didn't know where I can see the defaults!

I don't understand the difference between installing tsserver and null-ls? Do I need both?

Here's what I see when looking at the TypeScript file - in which gd, gr doesn't work – note that null-ls is `autostart: false`:

```
Language client log: /Users/akoskm/.local/state/nvim/lsp.log
 Detected filetype:   typescriptreact
 
 4 client(s) attached to this buffer: 
 
 Client: null-ls (id: 1, bufnr: [2, 4])
 	filetypes:       fish, lua, luau, sh, markdown, typescript, markdown.mdx, less, scss, css, vue, javascript, handlebars, jsonc, json, typescriptreact, graphql, yaml, html, javascriptreact
 	autostart:       false
 	root directory:  /Users/akoskm/Projects/my-project
 	cmd:             <function>
 
 Client: tailwindcss (id: 2, bufnr: [2, 4])
 	filetypes:       aspnetcorerazor, astro, astro-markdown, blade, clojure, django-html, htmldjango, edge, eelixir, elixir, ejs, erb, eruby, gohtml, gohtmltmpl, haml, handlebars, hbs, html, html-eex, heex, jade, leaf, liquid, mdx, mustache, njk, nunjucks, php, razor, slim, twig, css, less, postcss, sass, scss, stylus, sugarss, javascript, javascriptreact, reason, rescript, typescript, typescriptreact, vue, svelte
 	autostart:       true
 	root directory:  /Users/akoskm/Projects/my-project
 	cmd:             /Users/akoskm/.local/share/nvim/mason/bin/tailwindcss-language-server --stdio
 
 Client: eslint (id: 3, bufnr: [2, 4])
 	filetypes:       javascript, javascriptreact, javascript.jsx, typescript, typescriptreact, typescript.tsx, vue, svelte, astro
 	autostart:       true
 	root directory:  /Users/akoskm/Projects/my-project/frontend/packages/app-ecosystem
 	cmd:             /Users/akoskm/.local/share/nvim/mason/bin/vscode-eslint-language-server --stdio
 
 Client: tsserver (id: 4, bufnr: [2, 4])
 	filetypes:       javascript, javascriptreact, javascript.jsx, typescript, typescriptreact, typescript.tsx
 	autostart:       true
 	root directory:  /Users/akoskm/Projects/my-project
 	cmd:             /Users/akoskm/.local/share/nvim/mason/bin/typescript-language-server --stdio
 
 Configured servers list: lua_ls, eslint, tailwindcss, tsserver
```

**Reply by jyuan0** [reply]
Just a thought, maybe it's a memory limit?: https://github.com/microsoft/TypeScript/issues/44951

**Reply by jyuan0** [reply]
> I don't understand the difference between installing tsserver and null-ls? Do I need both?

You need tsserver for Typescript. I think null-ls/none-ls isn't enabled by default in LazyVim anymore, so I'm guessing some LazyVim extra or config you added is bringing it in. They serve different purposes and it's ok to have both.

**Reply by akoskm** [reply]
The last comment says tsserver now has unlimited memory if run under node. I wonder if that's the case on my end and where I can check if it's not running with the memory-limited version.

**Reply by jyuan0** [reply]
tsserver was installed with mason (via mason-lspconfig). Assuming you're on a *nix, everything should be under `~/.local/share/nvim/mason`.

**Reply by akoskm** [reply]
In `~/.local/share/nvim/mason/packages/typescript-language-server/mason-schemas/lsp.json` I see something like this:

```
"typescript.tsserver.maxTsServerMemory":{"scope":"window","markdownDescription":"%configuration.tsserver.maxTsServerMemory%","type":"number","default":3072}
```

Could this be the actual limit?

**Reply by jyuan0** [reply]
Yes, possibly. But just to be clear, the mem limit was just a guess on my part. I don't want to give you a false impression.

**Reply by akoskm** [reply]
I understand, thank you. The projects works fine in VSCode so this is most likely something around how tsserver is set up inside neovim. I'll give this configuration option a shot.

I was also looking at https://github.com/williamboman/mason.nvim hoping to find the above config file somewhere in the repo. But as I understand, it doesn't have any configuration stored and pulls the configs from the language servers, for TypeScript: https://github.com/typescript-language-server/typescript-language-server (I'm wondering if I'm changing the correct config file).

**EDIT**

Should I change/add this config value inside ~/.config/nvim/lua somewhere?

**Reply by jyuan0** [reply]
Do you mean the `maxTsServerMemory` option? If you're directly changing the file I think it should be in effect. You could also pass it through nvim-lspconfig init options, but I'd hack it to test first.

https://github.com/typescript-language-server/typescript-language-server/blob/master/docs/configuration.md#initializationoptions

**Reply by akoskm** [reply]
> Do you mean the maxTsServerMemory option?

Yes, this is what I did.

_Update_

`~/.local/share/nvim/mason/packages/typescript-language-server/mason-schemas/lsp.json` is probably not the right place to update the memory limit of tsserver. I updated it to 8GB, but opening the project in VSCode where with Vim bindings gd, and gr work, the most memory-intensive node process uses around 5GB of memory:

![image](https://github.com/LazyVim/LazyVim/assets/3111627/e10f62fd-7f53-4034-ac33-224354b27033)

However, while opening the same project in neovim, while the ts/js language server is being initialized, the most-memory hungry process disappears right around 4.5GB:

![image](https://github.com/LazyVim/LazyVim/assets/3111627/15f56055-698d-4af2-aa3c-d3500a0a0a99)

The process stays alive in VSCode for the whole time. In neovim it disappears so it most likely crashes.

**Reply by jyuan0** [reply]
Is opening in VSCode using the same server as opening in Neovim?

Seems like the setting isn't taking effect in neovim, but I can't think of where else you could set it except for the init option (not sure how to pass that though). This does seem like a mem limit though.

**Reply by jyuan0** [reply]
You could try a tsconfig.json in your project root: https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md#tsserver

**Reply by akoskm** [reply]
The process was terminated again:

![image](https://github.com/LazyVim/LazyVim/assets/3111627/39b8ff6b-0938-4b8c-965d-7f8f0a0f082b)

In VSCode it keeps running and takes up around 5GB memory:

![image](https://github.com/LazyVim/LazyVim/assets/3111627/8cb4357e-6b60-43fb-8266-c5a3208285cd)

Also, VSCode autocomplete didn't offer me any keys for "mem" so I suspect this can be configured through tsconfig:

![image](https://github.com/LazyVim/LazyVim/assets/3111627/f15b6e76-1b79-453d-84cf-15024aca0912)

***Update***

The only thing I could find in the TSConfig reference regarding memory was https://www.typescriptlang.org/tsconfig#disableSizeLimit but it didn't help.

Could be that the limit is on the `node` process?

**Reply by jyuan0** [reply]
You could try setting `maxTsServerMemory` in `init_options` using the example code here: https://github.com/LazyVim/LazyVim/discussions/2043

**Reply by akoskm** [reply]
🙌 Thank you for the link

node process sitting at ~5GB without crashes so far, gd, gr works 🥲

Following the link, this is the configuration that I ended up with:

```lua
return {
  "neovim/nvim-lspconfig",
  opts = {
    servers = {
      tsserver = {
        init_options = {
          maxTsServerMemory = 8192,
        },
      },
    },
  },
}
```

I'll keep testing this today.

Thanks @jyuan0 🙇‍♂️

**Reply by akoskm** [reply]
I couldn't find how to mark this as an answer, so I quoted you below. Thanks again!

**Comment by akoskm** [comment]
> You could try setting `maxTsServerMemory` in `init_options` using the example code here: #2043

@jyuan0 suggestion helped.

Here's the configuration I ended up with:

> 🙌 Thank you for the link
> 
> node process sitting at ~5GB without crashes so far, gd, gr works 🥲
> 
> Following the link, this is the configuration that I ended up with:
> 
> ```lua
> return {
>   "neovim/nvim-lspconfig",
>   opts = {
>     servers = {
>       tsserver = {
>         init_options = {
>           maxTsServerMemory = 8192,
>         },
>       },
>     },
>   },
> }
> ```
> 
> I'll keep testing this today.
> 
> Thanks @jyuan0 🙇‍♂️
